engine/src/savegame.cpp

```diff
@@ -174,6 +180,99 @@ void SaveFileCopy( const char *src, const char *dst )
     }
 }
 
+class Utf8Checker {
+public:
+    bool validUtf8(const std::vector<BYTE> &data) {
+        for (long unsigned int i = 0; i < data.size(); ++ i) {
+            // 0xxxxxxx
+            int bit1 = (data[i] >> 7) & 1;
+            if (bit1 == 0) continue;
+            // 110xxxxx 10xxxxxx
+            int bit2 = (data[i] >> 6) & 1;
+            if (bit2 == 0) return false;
+            // 11
+            int bit3 = (data[i] >> 5) & 1;
+            if (bit3 == 0) {
+                // 110xxxxx 10xxxxxx
+                if ((++ i) < data.size()) {
+                    if (is10x(data[i])) {
+                        continue;
+                    }
+                    return false;
+                } else {
+                    return false;
+                }
+            }
+            int bit4 = (data[i] >> 4) & 1;
+            if (bit4 == 0) {
+                // 1110xxxx 10xxxxxx 10xxxxxx
+                if (i + 2 < data.size()) {
+                    if (is10x(data[i + 1]) && is10x(data[i + 2])) {
+                        i += 2;
+                        continue;
+                    }
+                    return false;
+                } else {
+                    return false;
+                }
+            }
+            int bit5 = (data[i] >> 3) & 1;
+            if (bit5 == 1) return false;
+            if (i + 3 < data.size()) {
+                if (is10x(data[i + 1]) && is10x(data[i + 2]) && is10x(data[i + 3])) {
+                    i += 3;
+                    continue;
+                }
+                return false;
+            } else {
+                return false;
+            }
+        }
+        return true;
+    }
+private:
+    inline bool is10x(int a) {
+        int bit1 = (a >> 7) & 1;
+        int bit2 = (a >> 6) & 1;
+        return (bit1 == 1) && (bit2 == 0);
+    }
+};
+
+std::vector<BYTE> readFile(std::string filename)
+{
+    std::ifstream file(filename, std::ios::binary);
+    file.unsetf(std::ios::skipws);
+    std::streampos fileSize;
+
+    file.seekg(0, std::ios::end);
+    fileSize = file.tellg();
+    file.seekg(0, std::ios::beg);
+
+    std::vector<BYTE> vec;
+    vec.reserve(fileSize);
+
+    std::copy(std::istream_iterator<BYTE>(file),
+          std::istream_iterator<BYTE>(),
+          std::back_inserter(vec));
+
+    file.close();
+
+    return vec;
+}
+
+bool checkSaveGame(std::string filename)
+{
+    string path = GetSaveDir() + filename;
+    std::vector<BYTE> savegame = readFile(path);
+    Utf8Checker* check = new Utf8Checker;
```

Comment:
Can we use a smart pointer here? Such as `boost::shared_ptr` or similar?

Reply:
By the way, as you mentioned boost, could you investigate on the chance to use the Py check instead of this as an alternative like @ermo suggested? I've found a `PYTHONCALLBACK2` macro in `src/python/python_class.h` which should accept an argument (our filename complete with path) and return a value.

Reply:
That does look promising. I also see a `PYTHONCALLBACK` macro, which takes one string argument instead of two, and `virtual void callFunction (std::string str)`, under `PythonClass`. I think the latter might be the way to go, perhaps.

Reply:
Yes but the first argument is the module function, so we need another one to pass the game path, confrot these uses:
`PYTHONCALLBACK(void, this->self, "Execute");`
`PYTHONCALLBACK2(void, self, "UnPickle",s);`

Reply:
Gotcha

Reply:
...meanwhile I managed to test the opposite, exposing the C++ check to Python; I put it in the `Director` module hosting the savegame data related functions, and the assets like `computer_lib.py` can do this:
```
import Director
...
if not Director.isUtf8SaveGame(game):
```

Reply:
@P-D-E Cool. That sounds like the path of least resistance for moving forward.

Reply:
If I understood correctly, the other proposed solution would have involved Python code calling C++ code calling more Python code. I think this way is a lot better, because it's simpler.

Reply:
Correct indeed, unless the Python check code could be called both by the assets (directly instead of Py -> C++ -> Py) and by the engine (via boost's `call_method`), but this way is still to be experimented with.

Reply:
@P-D-E just a quick check whether you actually added the smart pointer as @stephengtuggy suggested?

Reply:
@ermo Not yet; if changing
```
    Utf8Checker* check = new Utf8Checker;
```
into
```
    std::unique_ptr<Utf8Checker> check(new Utf8Checker);
```
is okay for you all, I can do it now. It compiles and runs correctly.

Reply:
@P-D-E that would be preferable yes; we're moving towards using more smart pointer storage objects